runtime/lua/vim/pack.lua

```diff
@@ -674,23 +674,27 @@ function M.add(specs, opts)
     install_list(plugs_to_install)
   end
 
-  -- Register and `:packadd` those actually on disk
-  for _, p in ipairs(plugs) do
-    if p.info.installed then
-      pack_add(p, opts.load)
-    end
-  end
-
   -- Delay showing all errors to have "good" plugins added first
   local errors = {} --- @type string[]
   for _, p in ipairs(plugs_to_install) do
     if p.info.err ~= '' then
       errors[#errors + 1] = ('`%s`:\n%s'):format(p.spec.name, p.info.err)
     end
   end
+
+  -- Register and `:packadd` those actually on disk
+  for _, p in ipairs(plugs) do
+    if p.info.installed then
+      local ok, err = pcall(pack_add, p, opts.load)
+      if not ok then
+        errors[#errors + 1] = ('`%s`:\n%s'):format(p.spec.name, err)
+      end
+    end
+  end
```

Comment:
Oh, that's nice! Avoiding two cycles is even better here:

```suggestion
  -- Register and load those actually on disk while collecting errors
  -- Delay showing all errors to have "good" plugins added first
  local errors = {} --- @type string[]
  for _, p in ipairs(plugs) do
    if p.info.installed then
      local ok, err = pcall(pack_add, p, opts.load) --[[@as string]]
      if not ok then
        p.info.err = err
      end
    end
    if p.info.err ~= '' then
      errors[#errors + 1] = ('`%s`:\n%s'):format(p.spec.name, p.info.err)
    end
  end
```

Reply:
Sorry, this diff doesn't work because it will make `vim.pack` not report errors when _installing_ a plugin.

For example, in this script, I miswrite `nvim-treesitter` as `nvimtreesitter`, so it cannot install the plugin, yet it doesn't report any errors.
```
vim.opt.packpath:append(vim.fn.stdpath('data') .. '/site')

local github = function(name)
	return 'https://github.com/' .. name
end

vim.pack.add {
	github 'nvimtreesitter/nvimtreesitter',
}
```
So please let me revert it

Reply:
Or another way would be making `install_list` filter packages that have `installed = false` by itself

Reply:
> Sorry, this diff doesn't work because it will make `vim.pack` not report errors when _installing_ a plugin.

Which one? The "one cycle" approach *does* report errors during installation. The `if p.info.err ~= '' then` is done outside of `if p.info.installed then` and it traverses installed plugins also. I just checked it.

Reply:
Oh, it does. Sorry, I should have run `messages` to check